feg/cloud/go/services/basic_acct/servicers/service.go

```diff
@@ -0,0 +1,232 @@
+
+	// notls flag, use to disable TLS
+	notlsFlag = flag.Bool("notls", false, `Disable TLS when set`)
+
+	// insecure flag, use to disable server TLS cert verification
+	insecureFlag = flag.Bool("insecure", false, `Disable TLS server certificate verification`)
+)
+
+// BaseAccService Configuration
+type Config struct {
+	RemoteAddr   string `yaml:"remote_address"`
+	RootCaCert   string `yaml:"root_ca_cert"`
+	ClientCrt    string `yaml:"client_cert"`
+	ClientCrtKey string `yaml:"client_cert_key"`
+	NoTls        bool   `yaml:"no_tls"`
+	Insecure     bool   `yaml:"insecure"`
+}
+
+// BaseAccService
+type BaseAccService struct {
+	sync.RWMutex
+	cfg        *Config
+	remoteConn *grpc.ClientConn
+}
+
+// GetConfig returns a copy of current server configs
+func (s *BaseAccService) GetConfig() *Config {
+	if s != nil {
+		s.RLock()
+		defer s.RUnlock()
+		if s.cfg != nil {
+			cfgCopy := *s.cfg
+			return &cfgCopy
+		}
+	}
+	return nil
+}
+
+// Set Config updates current server configs
+func (s *BaseAccService) SetConfig(cfg *Config) {
+	if s != nil && cfg != nil {
+		s.Lock()
+		s.cfg = cfg
+		s.Unlock()
+	}
+}
+
+// NewBaseAcctService creates a new BaseAccService and initializes its configs
+func NewBaseAcctService() *BaseAccService {
+	cfg := &Config{
+		RemoteAddr:   *remoteAddrFlag,
+		RootCaCert:   *caCertFlag,
+		ClientCrt:    *clientCertFlag,
+		ClientCrtKey: *clientKeyFlag,
+		NoTls:        *notlsFlag,
+		Insecure:     *insecureFlag,
+	}
+	_, _, err := config.GetStructuredServiceConfig(feg.ModuleName, basic_acct.ServiceName, cfg)
+	if err != nil {
+		glog.Warningf("Failed reading '%s' service config: %v; Using the following defaults: %+v",
+			basic_acct.ServiceName, err, *cfg)
+	}
+	return &BaseAccService{cfg: cfg}
+}
+
+// Start will be called at the end of every new user session creation
+// start is responsible for verification & initiation of an accounting contract
+// between the user identity provider/MNO and service provider (ISP/WISP/PLTE)
+// A non-error return will indicate successful contract establishment and will
+// result in the beginning of service for the user
+func (s *BaseAccService) Start(ctx context.Context, req *protos.AcctSession) (*protos.AcctSessionResp, error) {
+	provider, consumer, gw, err := RetrieveParticipants(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	client, outCtx, cancel, err := s.GetAcctClient()
+	if err != nil {
+		return nil, err
+	}
+	defer cancel()
+
+	anSession := ToAnAcctSession(req)
+	anSession.ConsumerId = consumer
+	anSession.ProviderId = provider
+	anSession.ProviderGatewayId = gw
+
+	resp, err := client.Start(outCtx, anSession)
+	if err != nil {
+		return nil, err
+	}
+	return &protos.AcctSessionResp{
+		ReportingAdvisory: &protos.AcctSessionResp_ReportLimits{
+			OctetsIn:       resp.GetReportingAdvisory().GetOctetsIn(),
+			OctetsOut:      resp.GetReportingAdvisory().GetOctetsOut(),
+			ElapsedTimeSec: resp.GetReportingAdvisory().GetElapsedTimeSec(),
```

Comment:
is this the total quota given to the user?

Reply:
It's mostly for the future use since neither CWAG nor AGW currently have means of enforcing it. When session is created, the smart contract accounting service can request that the update for this session should be issued as soon as new usage reaches either  OctetsIn, OctetsOut or after ElapsedTimeSec from previous update, whatever comes first.